### ISSUE TITLE
smart math($number_or_object) methods, or separate math(self $object) and scalerMath($scaler) methods

### ISSUE THREAD

**Issue by Beakerboy** [markrogoyski/math-php#220]
How do you prefer classes define math operations? Should the class be flexible in what input is allowed, or should we type hint separate methods? The Polynomial class type hints Polynomial::add(Polynomial $poly). In order to add a scaler to a polynomial, it has to be converted into an object first. Adding a scalerAdd($number) method would avoid this.

Alternatively, the Polynomial::add() method could be altered to accept either is_numeric, or self objects.

Which do you prefer?

**Comment by markrogoyski** [comment]
I think the preference is be flexible and not force the user to know implementation details unless it is necessary.

For example in the Matrix class multiply and scalarMultiply are different functions accepting different types because they fundamentally do different things. For the Polynomial, any integer or float could be considered a zero-degree polynomial, so the operation is the same--only the input format is different. So in that case I'd accept both and perhaps just convert the number to a polynomial so there is only logic to do the addition in one place. So it would follow the Complex number class' example.

For the namespace, Complex number isn't a function, so it wouldn't go in there. A Polynomial is more an expression that represents a function, so it is a bit more than a number. What would the new shared namespace be? Or do you think putting in the Number namespace makes sense?

**Comment by Beakerboy** [comment]
I don't know WHAT a shared namespace would be called. The Arithmetic, Support, and Special classes are all libraries of static methods, while Piecewise and Polynomial are instantiable objects. In my mind, this makes them very different. All the other classes could be combined, rearranged, or what have you and, conceptually, everything is still the same. It just FEELS like polynomial and Piecewise have more in common with Complex than the rest in the current namespace.

**Comment by markrogoyski** [comment]
I agree. I originally made the Functions namespace for the Special functions. Then when the need for Map happened, Functions seems like the best place to put it. And it grew from there.

If there is no unifying Namespace, I think moving the expression classes like Polynomial to Number is fine. Up until now those classes did not have a published API (not listed in the main README), so it is not a big deal to move them.
